### PR TITLE
p2p: invert peer discovery integration

### PR DESCRIPTION
**Pull request by corverroos**
Inverts the integration between  discv5 and libp2p from pull to push.

category: refactor
ticket: #985 
feature_flag: invert_discv5

### CODE REVIEW

File: p2p/discovery.go [hunk 1]
```diff
@@ -16,24 +16,28 @@
...
+	"github.com/obolnetwork/charon/app/lifecycle"
+	"github.com/obolnetwork/charon/app/log"
 	"github.com/obolnetwork/charon/app/z"
 )
 
-// UDPNode wraps a discv5 udp node and adds the bootnodes relays.
-type UDPNode struct {
-	*discover.UDPv5
-	Relays []Peer
-}
```
Comment (corverroos) [inline p2p/discovery.go:35]: not used

File: p2p/gater_test.go [hunk 2]
```diff
@@ -71,7 +71,7 @@ func TestP2PConnGating(t *testing.T) {
 
 	err = nodeA.Connect(context.Background(), addr)
 	require.Error(t, err)
-	require.Contains(t, err.Error(), fmt.Sprintf("gater rejected connection with peer %s and addr %s", addr.ID, addr.Addrs[0]))
```
Comment (corverroos) [inline p2p/gater_test.go:74]: flapping test, somehow used ipv6 on my local machine

File: p2p/discovery.go [hunk 3]
```diff
@@ -133,3 +137,62 @@ func NewLocalEnode(config Config, key *ecdsa.PrivateKey) (*enode.LocalNode, *eno
 
 	return node, db, nil
 }
+
+// NewDiscoveryAdapter returns a life cycle hook that links discv5 to libp2p by
+// continuously polling discv5 for latest peer ERNs and adding then to libp2p peer store.
```
Comment (xenowits) [inline p2p/discovery.go:142]:
```suggestion
// continuously polling discv5 for latest peer ENRs and adding then to libp2p peer store.
```

File: p2p/discovery.go [hunk 4]
```diff
@@ -133,3 +137,62 @@ func NewLocalEnode(config Config, key *ecdsa.PrivateKey) (*enode.LocalNode, *eno
...
+func NewDiscoveryAdapter(tcpNode host.Host, udpNode *discover.UDPv5, peers []Peer) lifecycle.HookFuncCtx {
+	return func(ctx context.Context) {
+		if !featureset.Enabled(featureset.InvertDiscv5) {
+			return
+		}
+
+		ctx = log.WithTopic(ctx, "p2p")
+		ttl := peerstore.TempAddrTTL
+		baseDelay := expbackoff.WithBaseDelay(time.Millisecond * 100) // Poll quickly on startup
+		maxDelay := expbackoff.WithMaxDelay(ttl * 9 / 10)             // Slow down to 90% of ttl
```
Comment (xenowits) [inline p2p/discovery.go:152]: nit:
```suggestion
		maxDelay := expbackoff.WithMaxDelay(ttl * 0.9)             // Slow down to 90% of ttl
```

**Review by xenowits — APPROVED** [review]
Looks good!

File: p2p/discovery.go (hunk 4 above)
Comment (corverroos) [reply p2p/discovery.go:152]: doesn't compile

**Pull request merged**